### ISSUE TITLE
Do we need output-storage array inputs in `Scan` nodes?

### ISSUE THREAD

**Issue by brandonwillard** [aesara-devs/aesara#1232]
Do we really need the inputs that represent output storage in `Scan` nodes?  Is it possible that we could replace those inputs with just the initial values for the corresponding taps and leave it up to the implementations to handle how storage is handled?

Consider the following example:
```python
import aesara
import aesara.tensor as at


x0 = at.scalar("x0")
scan_out, _ = aesara.scan(lambda x_tm1: x_tm1 + 1, outputs_info={"initial": x0, "taps": [-1]}, n_steps=3)

aesara.dprint(scan_out, print_op_info=True)
# Subtensor{int64::} [id A]
#  |for{cpu,scan_fn} [id B] (outer_out_sit_sot-0)
#  | |TensorConstant{3} [id C] (n_steps)
#  | |IncSubtensor{Set;:int64:} [id D] (outer_in_sit_sot-0)
#  |   |AllocEmpty{dtype='float64'} [id E]
#  |   | |Elemwise{add,no_inplace} [id F]
#  |   |   |TensorConstant{3} [id C]
#  |   |   |Subtensor{int64} [id G]
#  |   |     |Shape [id H]
#  |   |     | |Unbroadcast{0} [id I]
#  |   |     |   |InplaceDimShuffle{x} [id J]
#  |   |     |     |x0 [id K]
#  |   |     |ScalarConstant{0} [id L]
#  |   |Unbroadcast{0} [id I]
#  |   |ScalarFromTensor [id M]
#  |     |Subtensor{int64} [id G]
#  |ScalarConstant{1} [id N]
# 
# Inner graphs:
# 
# for{cpu,scan_fn} [id B] (outer_out_sit_sot-0)
#  >Elemwise{add,no_inplace} [id O] (inner_out_sit_sot-0)
#  > |*0-<TensorType(float64, ())> [id P] -> [id D] (inner_in_sit_sot-0)
#  > |TensorConstant{1} [id Q]
```

In this example, the input labeled `id D` is the output-storage array where all the initial values (i.e. `x0`) and results from each iteration are stored.  Those values also comprise the one and only output of that `Scan` node (i.e. labeled `id B`).

By the time we compute the `Scan` node, that storage array has already been allocated, so, if we don't use it, it's a waste.  Also, because it's allocated in this way, the underlying memory won't be freed until the `VM` decides it can be freed; however, once the node is finished, that memory is no longer relevant and can be freed, but this design prevents the underling `Scan` implementation from managing that choice itself.  While this storage could be considered "fixed" and re-usable, it doesn't always make sense to force it at this level.

For instance, this approach seems to be part of a much bigger design issue: **`while`-loops require up-front allocations**.  It would be much better if we could "dynamically" allocate the output storage for a `while`-loop with an unknown number of iterations, and this output-storage design choice seems to work against that approach.
